ethdb/pebble/pebble.go

```diff
@@ -0,0 +1,576 @@
+	nonlevel0CompGauge  metrics.Gauge // Gauge for tracking the number of table compaction in non0 level
+	seekCompGauge       metrics.Gauge // Gauge for tracking the number of table compaction caused by read opt
+	manualMemAllocGauge metrics.Gauge // Gauge for tracking amount of non-managed memory currently allocated
+
+	quitLock sync.Mutex      // Mutex protecting the quit channel access
+	quitChan chan chan error // Quit channel to stop the metrics collection before closing the database
+
+	log log.Logger // Contextual logger tracking the database path
+
+	activeComp          int       // current number of active compactions
+	compStartTime       time.Time // the start time of the earliest currently-active compaction
+	compTime            int64     // total time spent in compaction in ns
+	level0Comp          uint32    // total number of level-zero compactions
+	nonLevel0Comp       uint32    // total number of non level-zero compactions
+	writeDelayStartTime time.Time // the start time of the latest write stall
+	writeDelayCount     int64     // total number of write stall counts
+	writeDelayTime      int64     // total time spent in write stalls
+}
+
+func (d *Database) onCompactionBegin(info pebble.CompactionInfo) {
+	if d.activeComp == 0 {
+		d.compStartTime = time.Now()
+	}
+	l0 := info.Input[0]
+	if l0.Level == 0 {
+		atomic.AddUint32(&d.level0Comp, 1)
+	} else {
+		atomic.AddUint32(&d.nonLevel0Comp, 1)
+	}
+	d.activeComp++
+}
+
+func (d *Database) onCompactionEnd(info pebble.CompactionInfo) {
+	if d.activeComp == 1 {
+		atomic.AddInt64(&d.compTime, int64(time.Since(d.compStartTime)))
+	} else if d.activeComp == 0 {
+		panic("should not happen")
+	}
+	d.activeComp--
+}
+
+func (d *Database) onWriteStallBegin(b pebble.WriteStallBeginInfo) {
+	d.writeDelayStartTime = time.Now()
+}
+
+func (d *Database) onWriteStallEnd() {
+	atomic.AddInt64(&d.writeDelayTime, int64(time.Since(d.writeDelayStartTime)))
+}
+
+// New returns a wrapped pebble DB object. The namespace is the prefix that the
+// metrics reporting should use for surfacing internal stats.
+func New(file string, cache int, handles int, namespace string, readonly bool) (*Database, error) {
+	var pebbleDb *Database
+	// Ensure we have some minimal caching and file guarantees
+	if cache < minCache {
+		cache = minCache
+	}
+	if handles < minHandles {
+		handles = minHandles
+	}
+	logger := log.New("database", file)
+	logger.Info("Allocated cache and file handles", "cache", common.StorageSize(cache*1024*1024), "handles", handles)
+
+	eventListener := &pebble.EventListener{
+		CompactionBegin: func(info pebble.CompactionInfo) {
+			pebbleDb.onCompactionBegin(info)
+		},
+		CompactionEnd: func(info pebble.CompactionInfo) {
+			pebbleDb.onCompactionEnd(info)
+		},
+		WriteStallBegin: func(info pebble.WriteStallBeginInfo) {
+			pebbleDb.onWriteStallBegin(info)
+		},
+		WriteStallEnd: func() {
+			pebbleDb.onWriteStallEnd()
+		},
+	}
```

Comment:
```go
&pebble.EventListener{
	CompactionBegin: pebbleDb.onCompactionBegin,
	CompactionEnd:   pebbleDb.onCompactionEnd,
	WriteStallBegin: pebbleDb.onWriteStallBegin,
	WriteStallEnd:   pebbleDb.onWriteStallEnd,
}
```

Reply:
Also, just inline this into the options declataion below, no need to have this as a separate variable really.